File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        result = ListNode(0, head)
        ptr_1 = result
        ptr_2 = head

        for _ in range(n):
            ptr_2 = ptr_2.next

        while ptr_2:
            ptr_1 = ptr_1.next
            ptr_2 = ptr_2.next

        ptr_1.next = ptr_1.next.next

        return result.next

File: Data Structures & Algorithms/reorder-linked-list/submission-1.kt
/**
 * Definition for singly-linked list.
d
 */

class Solution {
    fun reorderList(head: ListNode?): Unit {

        var slow: ListNode? = head
        var fast: ListNode? = head?.next

        while (fast != null && fast.next != null) {
            slow = slow?.next
            fast = fast?.next?.next
        }

        val second = slow?.next
        slow?.next = null

        var previous: ListNode? = null
        var current = second

        while (current != null) {
            var tmp = current.next
            current.next = previous
            previous = current
            current = tmp
        }

        var firstList: ListNode? = head
        var secondList: ListNode? = previous

        while (firstList != null && secondList != null) {
            val tmp1 = firstList.next
            val tmp2 = secondList.next

            firstList.next = secondList
            secondList.next = tmp1

            firstList = tmp1
            secondList = tmp2
        }
    }
}
